### DISCUSSION TITLE
Integrate Twin + Next.js + Emotion with App Router

### DISCUSSION THREAD

**Discussion by daryldixon1994** [ben-rogerson/twin.macro#826] [General]
Hello community, I just wanted to ask how to use Twin + Next.js + Emotion in a next js app bur when I am using the app router. The documentations available are destinated for the Pages Router.  Please can any one help me

**Comment by ben-rogerson** [comment]
Hi there
I've just updated the next-emotion-typescript example to use the new app router.
The trick is to use the jsx pragma at the top of files using emotion, eg:

```js
/** @jsxImportSource @emotion/react */
```

and [remove the @babel/preset-react plugin](https://github.com/ben-rogerson/twin.examples/commit/351e4ad11d612666feb6b2886a9191b1e8b3771f#diff-8e7430aee7d110ee12e0366c43b2e8328e0eae8fe870a88eae4bbb7532ec26e1L30) that defines the pragma.

Ideally I'd like to avoid the pragma at the top of each file but that's the only solution I've found so far.

It's early days on this and I don't have instructions yet, but [here's the changes to upgrade from the pages dir to the app dir](https://github.com/ben-rogerson/twin.examples/commit/351e4ad11d612666feb6b2886a9191b1e8b3771f).

Let me know if you encounter any issues.
Cheers

**Reply by u3u** [reply]
Hello, perhaps you can set `"jsxImportSource": "@emotion/react"` in `tsconfig.json` instead of manually defining it in each file (I haven't tried it yet, but it has always been like this in my configuration). 🤔

**Reply by ben-rogerson** [reply]
Hi - yeah that didn't work for me. I think we'd need a solution that works outside of typescript - especially for those non-typescript setups.

**Reply by u3u** [reply]
https://nextjs.org/docs/architecture/nextjs-compiler#importsource
Outside of TypeScript, you can use `jsconfig.json` to configure it. Since I haven't tried the `app` directory yet, I cannot test it. However, I have removed the automatic import of `@babel/preset-react` in `with-twin`, and it still works.

By the way, do I have to migrate to the `app` directory when using Server Components? Can I use Server Components in the `pages` directory?

**Reply by ben-rogerson** [reply]
Interesting - I'll have to give it this another try. 

The server components features are tied to the `app` directory - you could test with [the emotion example I've already migrated](https://github.com/ben-rogerson/twin.examples/tree/master/next-emotion-typescript), just add preset-react back into `withTwin.js` and specify the jsx pragma in the tsconfig rather than each of the tsx files.

**Reply by u3u** [reply]
Thanks! The server component upgrade is too significant, which can greatly reduce the size of JS sent to the client to improve [PageSpeed Insights](https://pagespeed.web.dev/) performance score.

**Reply by u3u** [reply]
Hi, I found a [swc-useclient](https://github.com/coder-xiaotian/swc-useclient) plugin (by @coder-xiaotian), which should automatically add the `"use client"` directive at the top of files in a specified directory. It would be even better if it supports adding custom content at the top of files through options!

I have no research on Rust, and I don't really like its syntax, but it does look very concise. The plugin is also implemented with only a few lines of code. 👍

**Reply by daryldixon1994** [reply]
Hello and thank you for your answers, but first of all should I have to follow the instructions here[https://github.com/ben-rogerson/twin.examples/tree/master/next-emotion](url) and then try your sollution above ? or @ben-rogerson could you please show me the ordered steps to do it , so I can integrate twin macro in my next js (app dir) project note that I am using JavasScript not TypeScript, because I am feeling that things are getting complicated and I don't know what t do.

**Reply by daryldixon1994** [reply]
in this docs https://github.com/ben-rogerson/twin.examples/tree/master/next-emotion, I have to import the GlobalStyles file in src/pages/_app.js like this: 

import GlobalStyles from '../styles/GlobalStyles'

const App = ({ Component, pageProps }) => (
  <>

    <GlobalStyles />
    <Component {...pageProps} />
  </>
)

export default App

this is the case of pager dir, but in the case of app dir where should I put these line of code, in the app directory I have only two file page.js or layout.js and both of them are not working and showed me the following error : 

`- error ./node_modules/cosmiconfig/dist/readFile.js:9:0
Module not found: Can't resolve 'fs'

https://nextjs.org/docs/messages/module-not-found

Import trace for requested module:
./node_modules/cosmiconfig/dist/Explorer.js
./node_modules/cosmiconfig/dist/index.js
./node_modules/babel-plugin-macros/dist/index.js
./node_modules/twin.macro/macro.js
./src/styles/GlobalStyles.js
./src/app/layout.js`

**Reply by coder-xiaotian** [reply]
@u3u Thanks for your affirmation! I have released a new version of use-client, now you can insert custom content at the top of files. You can have a try.🤝
<img width="869" alt="image" src="https://github.com/ben-rogerson/twin.macro/assets/22257014/5267b8dc-e723-4b8b-b926-ab4bf5f8f535">
